nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py

```diff
@@ -0,0 +1,255 @@
+                layer.activations_checkpoint_method = (
+                    None  # (@adithyare) adapter learning does not support activations checkpointing atm.
+                )
+
+        logging.info(f'Before adding adapters:\n{self.frozen_model.summarize()}')
+        self.frozen_model.freeze()
+        for _, module in self.frozen_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    module.add_adapter(
+                        name=adapter_key,
+                        cfg=ParallelLinearAdapterConfig(
+                            in_features=frozen_model_cfg.hidden_size,
+                            dim=cfg.adapter_tuning.adapter_dim,
+                            norm_position=cfg.adapter_tuning.norm_position,
+                            dropout=cfg.adapter_tuning.adapter_dropout,
+                        ),
+                    )
+
+        logging.info(f'After adding adapters:\n{self.frozen_model.summarize()}')
+
+    @classmethod
+    def list_available_models(cls):
+        pass
+
+    def forward(
+        self,
+        input_ids,
+        position_ids,
+        attention_mask,
+        taskname_ids,
+        labels=None,
+        inference=True,
+        set_inference_key_value_memory=False,
+        inference_max_sequence_len=None,
+    ):
+        if self.autocast_dtype == torch.float32:
+            output = self.frozen_model.model(
+                input_ids=input_ids,
+                position_ids=position_ids,
+                encoder_input=None,
+                attention_mask=attention_mask,
+                labels=labels,
+                set_inference_key_value_memory=set_inference_key_value_memory,
+                inference_max_sequence_len=inference_max_sequence_len,
+            )
+        else:
+            with torch.autocast(device_type="cuda", dtype=self.autocast_dtype):
+                output = self.frozen_model.model(
+                    input_ids=input_ids,
+                    position_ids=position_ids,
+                    encoder_input=None,
+                    attention_mask=attention_mask,
+                    labels=labels,
+                    set_inference_key_value_memory=set_inference_key_value_memory,
+                    inference_max_sequence_len=inference_max_sequence_len,
+                )
+
+        return output
+
+    def setup(self, stage=None):
+        if stage == 'predict' or self.virtual_prompt_style == VirtualPromptStyle.INFERENCE:
+            self.frozen_model.freeze()
+            return
+
+        self.setup_test_data()
+        if stage == 'test':
+            return
+
+        self.setup_training_data()
+        self.setup_validation_data()
+        logging.info(f'setup completed:\n{self.frozen_model.summarize()}')
+
+    def on_train_end(self):
+        # Save the best nemo model
+        self.save_to(save_path=self.cfg.nemo_path)
```

Comment:
Does the `.nemo` saving work properly here with TP > 2 and PP > 2? With regular GPT/T5 models we don't save `nemo` on train end, but use the `megatron_ckpt_to_nemo.py` converter to later convert checkpoints to nemo.

Reply:
It does work here. Have managed to get results for tp=2 and tp=8. Also works for pp=2 on a toy model. I can do more intensive checks down the line.